### PR TITLE
[POAE7-2932] support convert Dictionary encoding vector to Flat encoding vector

### PR DESCRIPTION
**Pull request by harborn**

### CODE REVIEW

File: cpp/src/cider/exec/module/batch/ArrowABI.h [hunk 1]
```diff
@@ -65,6 +66,37 @@ struct ArrowArray {
   void (*release)(struct ArrowArray*) = nullptr;
   // Opaque producer-specific data
   void* private_data{nullptr};
+
+  std::string toString(std::string indent = "") const {
```
Comment (jikunshang) [inline cpp/src/cider/exec/module/batch/ArrowABI.h:70]: better not change this file, we can move this function to utils.

File: cpp/src/cider/exec/processor/StatelessProcessor.cpp [hunk 2]
```diff
@@ -38,7 +38,6 @@ void StatelessProcessor::getResult(struct ArrowArray& array, struct ArrowSchema&
   auto output_batch = runtime_context_->getOutputBatch();
   output_batch->move(schema, array);
   runtime_context_->resetBatch(context_->getAllocator());
-  return;
```
Comment (jikunshang) [inline cpp/src/cider/exec/processor/StatelessProcessor.cpp:41]: keep this?

File: cpp/src/cider/exec/nextgen/jitlib/base/JITFunction.h [hunk 3]
```diff
@@ -73,15 +73,9 @@ class JITFunction {
...
-  template <typename R, typename... Args>
-  auto getFunctionPointer() {
-    if constexpr (sizeof...(Args) > 0) {
-      using func_type = R (*)(Args...);
-      return reinterpret_cast<func_type>(getFunctionPointer());
-    } else {
-      using func_type = R (*)();
-      return reinterpret_cast<func_type>(getFunctionPointer());
-    }
+  template <typename R, typename... Args, typename F = R (*)(Args...)>
```
Comment (jikunshang) [inline cpp/src/cider/exec/nextgen/jitlib/base/JITFunction.h:76]: why need this change?

File: cpp/src/cider-velox/src/CiderPipelineOperator.cpp [hunk 4]
```diff
@@ -36,7 +37,45 @@ bool CiderPipelineOperator::needsInput() const {
...
+    } else {
+      children.emplace_back(std::move(vectorPtr));
+    }
+  }
+  return std::make_shared<RowVector>(
+      operatorCtx_->pool(), rowTypePtr, BufferPtr(nullptr), input->size(), children);
+}
+
 void CiderPipelineOperator::addInput(RowVectorPtr input) {
+  input = convertDictionaryToFlat(input);
```
Comment (Ma-Jian1) [inline cpp/src/cider-velox/src/CiderPipelineOperator.cpp:78]: wrap in gflags

File: cpp/src/cider/exec/module/batch/ArrowABI.h [hunk 5]
```diff
@@ -65,6 +66,37 @@ struct ArrowArray {
...
+    std::stringstream ss;
+    ss << indent << "{" << '\n';
+    indent.push_back(' ');
+    indent.push_back(' ');
+    ss << indent << "length = " << length << '\n';
+    ss << indent << "null_count = " << null_count << '\n';
+    ss << indent << "offset = " << offset << '\n';
+    ss << indent << "n_buffers = " << n_buffers << '\n';
+    for (int64_t i = 0; i < n_buffers; ++i) {
+      double* p = (double*)buffers[i];
```
Comment (Ma-Jian1) [inline cpp/src/cider/exec/module/batch/ArrowABI.h:80]: should be `uintptr_t`

File: cpp/src/cider/exec/module/batch/ArrowABI.h [hunk 6]
```diff
@@ -65,6 +66,37 @@ struct ArrowArray {
...
+      if (p == nullptr) {
+        ss << indent << "buffers[" << i << "] is nullptr\n";
+      } else {
+        ss << indent << "buffers[" << i << "] = " << *p << '\n';
```
Comment (Ma-Jian1) [inline cpp/src/cider/exec/module/batch/ArrowABI.h:84]: print as `0xAB`

File: cpp/src/cider/exec/module/batch/ArrowABI.h [hunk 7]
```diff
@@ -65,6 +66,37 @@ struct ArrowArray {
...
+    indent.push_back(' ');
+    indent.push_back(' ');
+    ss << indent << "length = " << length << '\n';
+    ss << indent << "null_count = " << null_count << '\n';
+    ss << indent << "offset = " << offset << '\n';
+    ss << indent << "n_buffers = " << n_buffers << '\n';
+    for (int64_t i = 0; i < n_buffers; ++i) {
+      double* p = (double*)buffers[i];
+      if (p == nullptr) {
+        ss << indent << "buffers[" << i << "] is nullptr\n";
```
Comment (Ma-Jian1) [inline cpp/src/cider/exec/module/batch/ArrowABI.h:82]: `buffer[0x0]`

File: cpp/src/cider/exec/module/batch/ArrowABI.h [hunk 8]
```diff
@@ -65,6 +66,37 @@ struct ArrowArray {
...
+        ss << indent << "buffers[" << i << "] = " << *p << '\n';
+      }
+    }
+    ss << indent << "n_children = " << n_children << '\n';
+    for (int64_t i = 0; i < n_children; ++i) {
+      struct ArrowArray* paa = children[i];
+      ss << paa->toString(indent);
+    }
+    if (dictionary != nullptr) {
+      ss << indent << "dictionary field is not null\n";
```
Comment (Ma-Jian1) [inline cpp/src/cider/exec/module/batch/ArrowABI.h:93]: this line can be removed

File: cpp/src/cider/exec/nextgen/jitlib/base/JITFunction.h [hunk 9]
```diff
@@ -73,15 +73,9 @@ class JITFunction {
...
+  F getFunctionPointer() {
+    return reinterpret_cast<F>(getFunctionPointer());
```
Comment (Ma-Jian1) [inline cpp/src/cider/exec/nextgen/jitlib/base/JITFunction.h:78]:
```cpp
 auto getFunctionPointer() {
      using func_type = R (*)(Args...);
      return reinterpret_cast<func_type>(getFunctionPointer());
  }
```
may be enough

**Review by Ma-Jian1 — DISMISSED** [review]

**Review by Ma-Jian1 — DISMISSED** [review]

**Review by jikunshang — APPROVED** [review]

**Review by jikunshang — APPROVED** [review]

**Review by Ma-Jian1 — APPROVED** [review]

**Pull request merged**